### PR TITLE
Add request matcher

### PR DESCRIPTION
**Pull request by joelwurtz**
See https://github.com/php-http/client-common/pull/13

### CODE REVIEW

File: src/RequestMatcher.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * Match a request.
+ *
+ * @author Joel Wurtz <joel.wurtz@gmail.com>
+ */
+interface RequestMatcher
+{
+    /**
+     * Decides whether the rule(s) implemented by the strategy matches the supplied request.
+     *
+     * Equivalent of https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/RequestMatcherInterface.php
```
Comment (sagikazarmark) [inline src/RequestMatcher.php]:
```
PSR-7 implementation of Symfony's RequestMatcher

@see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/RequestMatcherInterface.php
```

File: src/RequestMatcher/RegexRequestMatcher.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Match a request with a regex on the uri.
+ *
+ * @author Joel Wurtz <joel.wurtz@gmail.com>
+ */
+class RegexRequestMatcher implements RequestMatcher
+{
+    /** @var string Matching regex */
```
Comment (sagikazarmark) [inline src/RequestMatcher/RegexRequestMatcher.php]: Please expand this to multiline.

**Comment by sagikazarmark** [comment]
Other possible matchers to add:

- CallbackMatcher
- ChainMatcher (AND or OR logic?)

**Comment by sagikazarmark** [comment]
LGTM :+1:

**Comment by joelwurtz** [comment]
Other matchers for another PR when we need it :)

**Comment by dbu** [comment]
i like the symfony approach to this better, actually: https://github.com/symfony/symfony/blob/582f4753a343f230fbe18b4e9a0747d48351ddfb/src/Symfony/Component/HttpFoundation/RequestMatcher.php

if we have a separate matcher for each part of the request, we build huge object trees.
i don't know if anybody ported the symfony request matcher to PSR-7 yet, if not this library could be a good place for it. though it might be worth to have it stand alone as it could be useful even when not using anything else from PHP-HTTP.

**Comment by sagikazarmark** [comment]
I don't see a reason why the regex one could not live together with the one you linked.

The message library is not dedicated to PHP-HTTP packages (but rather to PSR-7 as a utility package), so IMO we can have it in here. There aren't any heavy dependencies. (BTW: We should ask the FIG if they want a http-message-utils repo, similarly to cache-utils.)

But I am not against separating it to be standalone. (One reason against it: makes it harder to use it INSIDE message. For example I would use it the Matching authentication)

**Comment by dbu** [comment]
agreed, lets keep it inside message then. good idea to ask fig - if we
contribute parts of this to their world we can then depend on those
parts from message without worries and still rely on it.

my unhappiness with the RegexRequestMatcher is that it is only matching
one thing, instead of ip, host, path, method and so on like the symfony
one. if we add the full matcher, the regex matcher becomes redundant.
the name is a bit unclear too, its a RegexUriRequestMatcher, not regex
for host, path or method...

**Comment by joelwurtz** [comment]
> my unhappiness with the RegexRequestMatcher is that it is only matching
one thing

It's just a first implementation to have a simple example, as i'm only interested in the interface for the moment, will certainly write other implementations at a latter time :)

It can be renamed if you want, also didn't want a too long name, but yeah RegexUriRequestMatcher may be cleaner.

**Pull request merged**